### DISCUSSION TITLE
How do I increase performance for real time graphs?

### DISCUSSION THREAD

**Discussion by Gamebuster19901** [jfree/jfreechart#287] [Q&A] [answered]
I'm having a bit of an issue with creating a smooth chart that updates at a constant rate.

I've tried calling `chart.setNotify()` on a timer to only update the chart every few milliseconds, but the issue still persists.

Video of the problem: https://www.youtube.com/watch?v=SgjcZP30r9M

This is my ApplicationFrame: https://github.com/Gamebuster19901/SpeedrunHeart/blob/02bfaee2a4dc13f45ba311124a44bbac592a2b61/src/main/java/com/gamebuster19901/heart/gui/Monitor.java

**Comment by trashgod** [comment, marked as answer]
Never sleep on the [_event dispatch thread_](https://docs.oracle.com/javase/tutorial/uiswing/concurrency/initial.html), especially in a Swing `Timer` listener. A 5 ms duty cycle is unsustainable, and the timer's thread gradually saturates. Use [`SwingWorker`](https://docs.oracle.com/javase/tutorial/uiswing/concurrency/worker.html), for [example](https://stackoverflow.com/a/13205322/230513), limited to 30 Hz, or JavaFX, up to 60 Hz.

**Reply by Gamebuster19901** [reply]
I also had to slow down the serial output of the device, as when it would output a value of '0', the rate at which it sent data was significantly faster, causing many more dataset updates.

Doing what you suggested, and adding a 3ms delay when outputting trough the serial monitor has resolved the issue: https://github.com/Gamebuster19901/SpeedrunHeart/commit/55393bac1c13cfb472d734083d0f245267a0f0d9
